net processing: Extract `addr` send functionality into MaybeSendAddr()

jnewbery opened the pull request:
This continues the work of moving application layer data into net_processing. It refactors `addr` send functionality into its own function `MaybeSendAddr()` and flattens/simplifies the code. Isolating and simplifying the addr handling code makes subsequent changes (which will move addr data and logic into net processing) easier to review.

This is a pure refactor. There are no functional changes.

For motivation of the project, see #19398.

jnewbery committed:
[net processing] Extract `addr` send functionality into MaybeSendAddr()

Reviewer hint: review with

 `git diff --color-moved=dimmed-zebra --ignore-all-space`

jnewbery committed:
[net processing] Refactor MaybeSendAddr()

Changes to make MaybeSendAddr simpler and easier to maintain/update:

- assert invariant that node.vAddrToSend.size() can never exceed
  MAX_ADDR_TO_SEND
- erase known addresses from vAddrToSend in one pass
- no check for (vAddr.size() >= MAX_ADDR_TO_SEND) during iteration,
  since vAddr can never exceed MAX_ADDR_TO_SEND.

jnewbery committed:
[net processing] Change MaybeSendAddr() to take a reference

Change name of CNode parameter to node now that it's no longer a
pointer.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -4365,6 +4368,68 @@ void PeerManagerImpl::MaybeSendPing(CNode& node_to, Peer& peer)
+    pto.m_next_addr_send = PoissonNextSend(current_time, AVG_ADDRESS_BROADCAST_INTERVAL);
+
+    assert(pto.m_addr_known);
+    assert(pto.vAddrToSend.size() <= MAX_ADDR_TO_SEND);
```
It's currently impossible for `vAddrToSend` to grow larger than `MAX_ADDR_TO_SEND` (observe that the only place where elements are added to `vAddrToSend` check the size first). However, it might be safer to do something like:

```suggestion
    if(pto.vAddrToSend.size() > MAX_ADDR_TO_SEND) vAddrToSend.resize(MAX_ADDR_TO_SEND);
```

I'm happy to go with whatever reviewers prefer.

ajtowns replied:
Could have an Assume instead of an assert, but better to error out than silently cover it up if mistakes are introduced elsewhere I think.

jnewbery replied:
Do you think it'd be useful to add a new macro `AssertAndRecover()`:

- in debug builds, this asserts (terminates) if the condition fails
- in release builds, this executes some recovery code and continues.

eg here the recovery code would be to clear the `vAddrToSend` vector and return from `MaybeSendAddr()`. Better to not relay addr records to a peer and continue operation than crash the node.

ajtowns replied:
Isn't `AssertAndRecover(condition, recovery)` the same as `if (!Assume(condition)) { recovery; }` ? (Err, at least if `Assume` was the identity function its comments claim it to be, rather than being cast to `(void)` in release builds)

Not sure the recovery code is needed here -- more than MAX_ADDR_TO_SEND would be a protocol violation, but shouldn't cause a crash or anything, I think?

jnewbery replied:
> Isn't AssertAndRecover(condition, recovery) the same as if (!Assume(condition)) { recovery; }

Ah interesting. I didn't realise that.

> Not sure the recovery code is needed here -- more than MAX_ADDR_TO_SEND would be a protocol violation, but shouldn't cause a crash or anything, I think?

I think it's probably better to skip sending the `addr` message than violate the p2p protocol.

jnewbery replied:
#21317 has now been merged so I've updated this to use the `if (!Assume(condition)) { recovery; }` pattern.

jnewbery commented:
Rebased on master to pick up fix for interface_zmq.py in #21008.

ajtowns commented on src/net_processing.cpp:
```diff
@@ -337,7 +337,7 @@ class PeerManagerImpl final : public PeerManager
 
     /** Send `addr` messages on a regular schedule*/
-    void MaybeSendAddr(CNode* pto);
+    void MaybeSendAddr(CNode& pto);
```
Maybe change this to `node` or `node_to` as per `MaybeSendPing`. It's no longer a `p` either way.

jnewbery replied:
Done

ajtowns commented on src/net_processing.cpp:
```diff
@@ -4395,45 +4395,39 @@ void PeerManagerImpl::MaybeSendAddr(CNode& pto)
-        // we only send the big addr message once
-        if (pto.vAddrToSend.capacity() > 40)
-            pto.vAddrToSend.shrink_to_fit();
+    assert(pto.m_addr_known);
```
Should move this to directly after the `RelayAddrsWithConn` check -- that's what makes it true, and you rely on `m_addr_known` being not null in the `!IBD` branch above as well.

jnewbery replied:
Done

ajtowns reviewed (dismissed):
Haven't thoroughly reviewed the "Refactor MaybeSendAddr" commit; seems like separating it into three or four commits would be clearer (early exit; invariants; skip vAddr check; one-pass erase)?

jnewbery commented on src/net.h:
```diff
@@ -546,8 +546,8 @@ class CNode
-    std::chrono::microseconds m_next_addr_send GUARDED_BY(cs_sendProcessing){0};
-    std::chrono::microseconds m_next_local_addr_send GUARDED_BY(cs_sendProcessing){0};
+    std::chrono::microseconds m_next_addr_send{0};
+    std::chrono::microseconds m_next_local_addr_send{0};
```
These members should never have been guarded by `cs_sendProcessing`, which really only exists at this point to make sure that multiple threads don't enter `SendMessages()` concurrently.

The annotation was over-zealously added in #13123. The author of that PR agrees that these annotations are unnecessary: https://github.com/bitcoin/bitcoin/pull/13123#issuecomment-647505130.

These variables are only ever read/written by the message handler thread inside `SendMessages()`, so it's safe for them to be unguarded. I can make them atomic if you think that would make this more obviously safe.

ajtowns replied:
"These variables are only ever read/written ... inside SendMessages" -- is **exactly** what GUARDED_BY(cs_sendProcessing) documents. The point of having the guard is to catch errors when we change the code in future and violate the assumptions we were relying on -- if the code compiles with the GUARDED_BY statements, it will always also compile correctly without them -- the trouble comes when people make changes later.

I don't know what @practicalswift thinks the annotation was a mistake; it's not. (It's a mistake the other addr-related member variables aren't guarded though)

Having things be atomic instead of guarded is only a good idea if they're actually accessed by multiple threads, and either they don't need to be synchronized with any other data, or the timing is too important for proper locks.

jnewbery replied:
> It's a mistake the other addr-related member variables aren't guarded though

Ok. I've moved all of the addr fields to be guarded by their own `m_addr_mutex` mutex. This seems much better than having data/logic that should be in net_processing being locked by a mutex in net.

jnewbery replied:
That proved more difficult than expected. I've reverted that change and have moved just  `m_next_addr_send` and `m_next_local_addr_send` to be guarded by the new `m_addr_mutex`. I'll move the rest of the addr fields under that mutex when they move into the `Peer` struct.

ajtowns replied:
I don't follow why you're changing those guards at all in this PR; you don't need to. See https://github.com/ajtowns/bitcoin/tree/202003-pr21236

I don't think it will make sense to lock `m_addr_known` and `vAddrToSend` with per-peer locks: when receiving an addr message we'll obtain the lock for the peer that sends the message, then attempt to obtain the lock for a peer we want to forward the address on to -- but there's no particular reason to expect those peers to always be in a particular order, so holding the sending peer's lock would create ordering violations, and releasing and reacquiring would be a bit ridiculous given that this is always happening in a single thread.

jnewbery replied:
Thanks! Totally agree that taking and releasing multiple locks would be ridiculous. I've taken your branch.

jnewbery replied:
This is still causing confusion for reviewers, so I've moved the fields to have their own lock that is only used in net_processing.

jnewbery committed:
[net processing] Refactor MaybeSendAddr() - early exits

Add early exit guard clauses if node.RelayAddrsWithConn() is false or if
current_time < node.m_next_addr_send. Add comments.

This commit leaves some lines over-indented. Those will be fixed in a
subsequent whitespace-only commit.

jnewbery committed:
[net processing] Fix overindentation in MaybeSendAddr()

Reviewer hint: review with `git diff --ignore-all-space`.

jnewbery commented:
Thanks for the review @ajtowns. I've addressed your comments.

jnewbery commented:
Oops. I missed taking the new lock in one of the call sites. Will fix up tomorrow.

jnewbery commented:
Once all of the addr data has been moved into net_processing, I intend to guard it with a single lock (see the `m_addr_relay` struct in https://github.com/jnewbery/bitcoin/tree/2021-02-lazy-init-peer). However, it's proving difficult to get lock orders right while the data is still in net. I've decided to punt on doing that until after this PR.

To avoid simply removing the guard from `m_next_addr_send` and `m_next_local_addr_send`, I've introduced the new `m_addr_mutex` lock in this PR. Once the data is moved into `Peer`, I'll extend that mutex to guard the remaining addr data.

fanquake reviewed (commented):
Concept ACK - this seems like a fairly good simplification. I like consolidating logic and returning early when possible. If this SendMessages() logic is going to be split up even more, you could consider a call to GetTime<>() at the start, and just pass that into `MaybeSendPing`, `MaybeSendAddr`, `MaybeSomeOtherFunc`  etc.

This doesn't currently compile:
```bash
net_processing.cpp:4408:91: error: no member named 'm_addr_mutex' in 'CNode'
    auto addr_already_known = [&node](const CAddress& addr) EXCLUSIVE_LOCKS_REQUIRED(node.m_addr_mutex)
                                                                                     ~~~~ ^
./threadsafety.h:31:79: note: expanded from macro 'EXCLUSIVE_LOCKS_REQUIRED'
#define EXCLUSIVE_LOCKS_REQUIRED(...) __attribute__((exclusive_locks_required(__VA_ARGS__)))
                                                                              ^~~~~~~~~~~
1 error generated.
```

jnewbery committed:
[net processing] Only call GetTime() once in SendMessages()

We currently call GetTime() 4 times in SendMessages(). Consolidate this to
once GetTime() call.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -4366,6 +4369,71 @@ void PeerManagerImpl::MaybeSendPing(CNode& node_to, Peer& peer)
+
+    assert(node.m_addr_known);
+
+    auto current_time = GetTime<std::chrono::microseconds>();
```
done

jnewbery commented on src/net_processing.cpp:
```diff
@@ -4366,6 +4369,71 @@ void PeerManagerImpl::MaybeSendPing(CNode& node_to, Peer& peer)
+    node.vAddrToSend.clear();
+
+    // we only send the big addr message once
+    if (node.vAddrToSend.capacity() > 40)
+        node.vAddrToSend.shrink_to_fit();
```
done

jnewbery commented:
Thanks for the review, @fanquake! The compile error was due to an annotation that I'd left from a previous branch. It's now fixed.

> If this SendMessages() logic is going to be split up even more, you could consider a call to GetTime<>() at the start, and just pass that into MaybeSendPing, MaybeSendAddr, MaybeSomeOtherFunc etc.

Good idea. Done!

amitiuttarwar commented on src/net_processing.cpp:
```diff
@@ -4375,8 +4375,6 @@ void PeerManagerImpl::MaybeSendAddr(CNode& node, std::chrono::microseconds curre
-    const CNetMsgMaker msgMaker(node.GetCommonVersion());
-
     // Periodically advertise our local address to the peer.
     if (fListen && !::ChainstateActive().IsInitialBlockDownload() &&
```
might be worth updating to `m_chainman.ActivateChainstate()`? 

h/t https://github.com/bitcoin/bitcoin/pull/21327#discussion_r585433607

jnewbery replied:
This isn't adding a new `::ChainstateActive()` like #21327 was. I think changing this call to be `m_chainman.ActiveChainstate()` would be adding more noise/distraction for reviewers.

amitiuttarwar replied:
ok. since this is a refactor PR anyway, as a reviewer I would not mind, but its no worries :)

amitiuttarwar reviewed (commented):
approach ACK

I have some more things to think about / investigate (namely the lock changes & the `vAddrToSend` size assertion) , but left a few comments in my first pass.

amitiuttarwar commented on src/net_processing.cpp:
```diff
@@ -334,7 +334,10 @@ class PeerManagerImpl final : public PeerManager
     /** Send a ping message every PING_INTERVAL or if requested via RPC. May
      *  mark the peer to be disconnected if a ping has timed out. */
-    void MaybeSendPing(CNode& node_to, Peer& peer);
+    void MaybeSendPing(CNode& node_to, Peer& peer, std::chrono::microseconds now);
```
looks like both `MaybeSendPing` and `MaybeSendAddr` can be made into const member functions

jnewbery replied:
I'm not sure that `const` should be used in cases like this. The compiler will make sure that any function marked `const` doesn't mutate the object's data members, which is the case for both of these functions. However, `const` is not transitive, so the functions are still able to modify state through pointers and modify global state.

Here, `PeerManagerImpl` owns the map of `Peer` objects:

```
class PeerManagerImpl final : public PeerManager
{
[...]
private:
    /**
     * Map of all Peer objects, keyed by peer id. This map is protected
     * by the m_peer_mutex. Once a shared pointer reference is
     * taken, the lock may be released. Individual fields are protected by
     * their own locks.
     */
    std::map<NodeId, PeerRef> m_peer_map GUARDED_BY(m_peer_mutex);
```

and the map of `CNodeState` objects are global (in the file's unnamed namespace):

```
/** Map maintaining per-node state. */
static std::map<NodeId, CNodeState> mapNodeState GUARDED_BY(cs_main);
```

So `PeerManagerImpl` definitely owns the `Peer` objects, and logically owns the `CNodeState` objects since it's responsible for constructing and destructing them (and #20758 actually moves the `CNodeState` objects into `PeerManagerImpl`).

These functions don't mutate `PeerManagerImpl`, but _do_ mutate data that it owns. I think in such cases there's an argument that we shouldn't mark the function `const` even if the compiler would allow us to. For me, `const` on a member function is telling the programmer "this function does not mutate the logical state of this object", which isn't the case here.

ajtowns replied:
Since `PeerRef` is a shared_ptr which does its own destruction once the ref count hits zero, I don't think it's accurate to say that `PeerManagerImpl` owns the `Peer` object.

jnewbery replied:
I'm not sure it's right to say that ownership of Peer is shared because it uses a shared pointer. We use a shared pointer for the Peer for convenience/safety/synchronization. We never take a shared pointer to a Peer outside PeerManagerImpl, and if we really wanted to, could implement it with raw pointers and a manual ref count, and then `new` and `delete` the Peer objects (as is done by CConnman for its CNode objects in vNodes)

That's maybe a bit pedantic. I think the more important question is "what is `const` for?" For me, it means "this function won't mutate the _logical_ state of this object". If we're using it to say "this function won't mutate the data members of this object, but it will mutate the state of data structures that are integral to how this object operates, but will do so through a pointer", then it seems a bit legalistic and not very useful to me.

ajtowns replied:
I was surprised to discover that if you have a class Y with a member `X& x` then C++ thinks it's fine to modify `x.foo = 3` from a const member function of Y. So not sure I'd argue against "a bit legalistic and not very useful"... (So in particular, if you have a `const PeerManagerImpl`, it will have no problems calling non-const functions of its `m_connman`)

If you consider modifying the contents of a `PeerRef` to be modifying a `PeerManagerImpl`, then I don't think it's consistent to mark `GetPeerRef()` as const either (code compiles fine that way though). I dunno; "shared" makes me think things are shared, not owned, so that's definitely a mental adjustment.

I guess I don't think it's a good idea to try to put too much convention on top of what our compilers will actually enforce here though.

jnewbery replied:
This certainly isn't a hill I'm prepared to die on (either here or in https://github.com/bitcoin/bitcoin/pull/21162#discussion_r587410937). My preference is not to mark functions as const if they're mutating data that is so closely associated with the object that it can only be modified through the object, but I'll go along with the majority if people want this to be const.

ajtowns replied:
Yeah, not my hill either, but I think it's worth discussing so we've got a *shared* understanding of how this data's meant to be controlled... I have some other ideas about `Peer` ownership I want to try out once more of this refactoring is finished that might change this, and worst case can add or remove const markers when we're reordering the PeerManageImpl declarations and adding more comments...

amitiuttarwar replied:
I agree that the relevant question is "what is `const` for?" The two main aspects are compiler enforcement & human understanding. @jnewbery I get what you're saying about making these member functions `const` is misleading for the humans, because the function often mutates the logical state of the object. However, I also agree with @ajtowns that since the compilers would actually be enforcing something here, it does offer a bit of literal guarding?

I don't have a strong opinion on which way we go, but it is definitely an interesting consideration & mostly I conclude that `const` is complex. I'm personally curious to learn more about how a compiler uses the keyword- when its just enforcement vs optimization. (recommending resources welcome!)

sipa replied:
I'm 99% sure that compilers don't use `const` markers on functions/pointers/references at all for optimization. All they do is disallow certain operations on them.

It's different if you have an actual `const` variable (e.g. a global, local, or member variable that is declared const where it is defined & initialized), because that means "changing this is UB". Any other use simply means "no non-const operations are allowed on object *through this reference/pointer/function*; if other code is invoked that happens to hold a non-const reference/pointer, the object can still change (and the compiler can't assume a cached version in a register won't change e.g.).

jnewbery replied:
@sipa yes, that's exactly my understanding. Mutating a `const`-declared variable is UB. Therefore compilers could make optimizations based on the fact that `const`-declared variables cannot change.

`const` on a member functions simply means that this function call won't mutate the state of the object, although of course, other code could mutate the state. Therefore, the compiler can't make any assumptions about the state of the variable not changing.

You can even mutate the state through a `const` function by casting away the constness, eg:

```c++
#include <iostream>

class thing
{
    int x{0};

public:
    const void mutate(int x_in)
    {
        thing* mutable_this = const_cast<thing*>(this);
        mutable_this->x = x_in;
    }

    const int access() { return x; }

    thing(int x_in) : x(x_in) { }
};
 
int main()
{
    thing t(1);
    std::cout << "t.x = " << t.access() << "\n";

    t.mutate(2);
    std::cout << "t.x = " << t.access() << "\n";

    return 0;
}
```

You can see that the state of `t` has been mutated in the `mutate()` function, even though that function is `const`. This compiles with no warnings. I believe this is perfectly legal (although ill-advised!) c++. The c++ core guidelines advise against casting away const in almost all cases: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es50-dont-cast-away-const. Scott Meyers does use casting away const to avoid code duplication: https://stackoverflow.com/a/123995/933705 (although that book is now quite old, so I'm not sure if the advice is still current).

Going back to the main question about whether these functions _should_ be `const`, I found [the isocpp.org wikipedia page on const correctness](https://isocpp.org/wiki/faq/const-correctness) to be very useful in clarifying my thoughts. A few extracts:

> The trailing `const` on `inspect()` member function should be used to mean the method won’t change the object’s _abstract_ (client-visible) state. That is slightly different from saying the method won’t change the “raw bits” of the object’s `struct`

> a `const` member function must not change (or allow a caller to change) the this object’s _logical_ state (AKA _abstract_ state AKA _meaningwise_ state). Think of what an object means, not how it is internally implemented. A Person’s age and name are logically part of the Person, but the Person’s neighbor and employer are not. An inspector method that returns part of the this object’s _logical_ / _abstract_ / _meaningwise_ state must not return a non-`const` pointer (or reference) to that part, **independent of whether that part is internally implemented as a direct data-member physically embedded within the this object or some other way**. (emphasis mine)

> - If a method changes any part of the object’s logical state, it logically is a mutator; it should not be const even if (as actually happens!) the method doesn’t change any physical bits of the object’s concrete state.
> - Conversely, a method is logically an inspector and should be const if it never changes any part of the object’s logical state, even if (as actually happens!) the method changes physical bits of the object’s concrete state.

> If you are outside the class — you are a normal user, every experiment you could perform (every method or sequence of methods you call) would have the same results (same return values, same exceptions or lack of exceptions) irrespective of whether you first called that lookup method. If the lookup function changed any future behavior of any future method (not just making it faster but changed the outcome, changed the return value, changed the exception), then the lookup method changed the object’s logical state — it is a mutuator.

Based on that, I believe these functions should not be marked `const`. Calling `MaybeSendPing()` and `MaybeSendAddr()` multiple times will result in different behaviour that is observable from outside the class, since the `Peer()` objects are mutated.

amitiuttarwar replied:
I'm convinced :)

jnewbery commented on src/net_processing.cpp:
```diff
@@ -336,6 +336,9 @@ class PeerManagerImpl final : public PeerManager
      *  mark the peer to be disconnected if a ping has timed out. */
     void MaybeSendPing(CNode& node_to, Peer& peer, std::chrono::microseconds now);
 
+    /** Send `addr` messages on a regular schedule*/
```
Fixed.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -4365,6 +4368,71 @@ void PeerManagerImpl::MaybeSendPing(CNode& node_to, Peer& peer, std::chrono::mic
 }
 
+void PeerManagerImpl::MaybeSendAddr(CNode* pto, std::chrono::microseconds current_time)
+{
```
done

jnewbery commented on src/net_processing.cpp:
```diff
@@ -4368,68 +4368,67 @@ void PeerManagerImpl::MaybeSendPing(CNode& node_to, Peer& peer, std::chrono::mic
-        !::ChainstateActive().IsInitialBlockDownload() &&
-        pto->m_next_local_addr_send < current_time) {
+    if (fListen && node.RelayAddrsWithConn() &&
+        node.m_next_local_addr_send < current_time) {
```
Oops. Good catch. Fixed!

jnewbery commented:
Thanks for the review @amitiuttarwar! I've addressed all of your comments.

jnewbery commented:
Rebased

amitiuttarwar reviewed (commented):
code review ACK 151597e19b

a couple review notes
- I liked the use of the erase-remove idiom, my understanding of the changes: we no longer copy all the elements of `vAddrToSend` and instead are able to edit in place with linear complexity by using this pattern. I think we now traverse the vector 2x instead of 1x, but that seems fine. I also find the logic flow much easier to understand, esp with removing the unused logic to handle >1000 addrs queued. 
- the biggest change seems to be breaking this `addr` functionality out of `cs_main`. I spent some time checking if this is safe. to do so, I checked that all of the state touched either takes `cs_main` itself, is protected by a more specific lock, or only accessed from the message handler thread (eg. `m_addr_known`, `vAddrToSend`, `GetLocalAddrForPeer`..) 

~The only touchpoint I'm unsure about is the call to `fRelay` [here](https://github.com/bitcoin/bitcoin/blob/master/src/qt/optionsmodel.cpp#L499), I'm not very familiar with the qt code so I'm not 100% sure how this interacts with the rest. I think it's ok, but just wanted to explicitly call out the gap in my understanding.~

jnewbery commented on src/net_processing.cpp:
```diff
@@ -4196,8 +4194,20 @@ void PeerManagerImpl::MaybeSendAddr(CNode& node, std::chrono::microseconds curre
+
+    // Remove addr records that the peer already knows about
+    auto addr_already_known = [&node](const CAddress& addr) {return node.m_addr_known->contains(addr.GetKey());};
+    node.vAddrToSend.erase(std::remove_if(node.vAddrToSend.begin(), node.vAddrToSend.end(), addr_already_known),
```
Yes! Thank you :)

jnewbery replied:
Done!

jnewbery commented:
> The only touchpoint I'm unsure about is the call to fRelay here

I'm not sure what this is referring to. I don't see `fRelay` in that file at all.

jnewbery commented:
> I think we now traverse the vector 2x instead of 1x, but that seems fine.

I've changed this so we both check for the addr in `node.m_addr_known` and add it to `node.m_addr_known` in the same pass.

xzavjierr commented:
Well just depends if your locating a delivery or life form suppose time and
reality and computer s and virtually augmentation with magnetic field and
radio station tower in unison league s of things networked in a branch of
servers and databases uploading instead of down and cloud space could be
used as an example of connections that don't exist like bluetooth or sound
then theory has it if you left side of your brain was the tracking
transition of nuero synaptic feeds then suppose I could run all three bases
when not slowed down or placed in a geo metric analysis predicted spatial
of surveying that layers beyond the scope of even einstiens guessing or
math if I stepped left twice jumped stretched my calf and verily looked
right to check the sun then my own body would be in perfect balance in
relationship to this earth 🌎 and mutually connected to flow to and from
various spaces or matters that may not even be existing considering even
grass doesn't make since to most unless your in paradise bloom berg point
is two planets almost identical to another so close to one another that
some one could climb or fly or trampoline right into another world where
they might run into a duplicate smarter version of there very own self
suppose what I said is right in the sky now and the blue you see is the
water from another earth so close that account for pipelines and quotes of
gravity and you got a sky network of endless resources unless someone knew
more than what he just said just now

On Sat, Mar 6, 2021, 9:44 AM John Newbery <notifications@github.com> wrote:

> I think we now traverse the vector 2x instead of 1x, but that seems fine.
>
> I've changed this so we both check for the addr in node.m_addr_known and
> add it to node.m_addr_known in the same pass.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/21236#issuecomment-791986604>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ATDTIGQTWU7VXE4PDZOB63LTCJLWHANCNFSM4X4K2C3Q>
> .
>

amitiuttarwar reviewed (commented):
>> The only touchpoint I'm unsure about is the call to fRelay here

> I'm not sure what this is referring to. I don't see fRelay in that file at all.

oops, I'm not sure what I was referring to either 😳 
I've re-checked all the `fRelay` touch points and am now comfortable with pulling it out of `cs_main`. 

reACK c3b424d98f

jnewbery commented:
Rebased

jnewbery commented:
rebased

amitiuttarwar reviewed (approved):
reACK e3d6cbca83300e5ccc0fd8a1075ff52654aa76ff, rebased chainman & std::optional changes since last round.

hebasto commented:
Concept ACK.

hebasto reviewed (commented):
Approach ACK e3d6cbca83300e5ccc0fd8a1075ff52654aa76ff, ~need more time to think about the last commit~.

1. The message of the commit "[net processing] Only call GetTime() once in SendMessages()" (461dda316aefa4f792350a82ca2fb8c6daf8cca1) states:
> We currently call GetTime() 3 times in SendMessages(). Consolidate this to
once GetTime() call.

I see two calls of three are consolidated. What about the third one:
https://github.com/bitcoin/bitcoin/blob/e3d6cbca83300e5ccc0fd8a1075ff52654aa76ff/src/net_processing.cpp#L4513?

Also I believe that this commit is not a clean refactoring. To be precise, change behavior could be expected as even `LOCK(cs_main);` could take unpredictable time. But the scale of such changes seems insignificant to P2P, right?

2. The commit "[net processing] Extract `addr` send functionality into MaybeSendAddr()" (97e101345e6b838a5dd308a6adafea53eb14d521) allows to localize acquiring of the `CNode::cs_sendProcessing` mutex:
```diff
--- a/src/net.cpp
+++ b/src/net.cpp
@@ -2164,10 +2164,7 @@ void CConnman::ThreadMessageHandler()
             if (flagInterruptMsgProc)
                 return;
             // Send messages
-            {
-                LOCK(pnode->cs_sendProcessing);
-                m_msgproc->SendMessages(pnode);
-            }
+            m_msgproc->SendMessages(pnode);
 
             if (flagInterruptMsgProc)
                 return;
diff --git a/src/net.h b/src/net.h
index 48d37084a..300c9fffe 100644
--- a/src/net.h
+++ b/src/net.h
@@ -419,7 +419,7 @@ public:
     std::list<CNetMessage> vProcessMsg GUARDED_BY(cs_vProcessMsg);
     size_t nProcessQueueSize{0};
 
-    RecursiveMutex cs_sendProcessing;
+    Mutex cs_sendProcessing;
 
     uint64_t nRecvBytes GUARDED_BY(cs_vRecv){0};
 
@@ -787,7 +787,7 @@ public:
     * @param[in]   pnode           The node which we are sending messages to.
     * @return                      True if there is more work to be done
     */
-    virtual bool SendMessages(CNode* pnode) EXCLUSIVE_LOCKS_REQUIRED(pnode->cs_sendProcessing) = 0;
+    virtual bool SendMessages(CNode* pnode) = 0;
 
 
 protected:
diff --git a/src/net_processing.cpp b/src/net_processing.cpp
index 8b13a859c..b7c6a08ea 100644
--- a/src/net_processing.cpp
+++ b/src/net_processing.cpp
@@ -241,7 +241,7 @@ public:
     void InitializeNode(CNode* pnode) override;
     void FinalizeNode(const CNode& node, bool& fUpdateConnectionTime) override;
     bool ProcessMessages(CNode* pfrom, std::atomic<bool>& interrupt) override;
-    bool SendMessages(CNode* pto) override EXCLUSIVE_LOCKS_REQUIRED(pto->cs_sendProcessing);
+    bool SendMessages(CNode* pto) override;
 
     /** Implement PeerManager */
     void CheckForStaleTipAndEvictPeers() override;
@@ -323,7 +323,7 @@ private:
     void MaybeSendPing(CNode& node_to, Peer& peer, std::chrono::microseconds now);
 
     /** Send `addr` messages on a regular schedule. */
-    void MaybeSendAddr(CNode& node, std::chrono::microseconds current_time) EXCLUSIVE_LOCKS_REQUIRED(node.cs_sendProcessing);
+    void MaybeSendAddr(CNode& node, std::chrono::microseconds current_time);
 
     const CChainParams& m_chainparams;
     CConnman& m_connman;
@@ -4160,7 +4160,7 @@ void PeerManagerImpl::MaybeSendPing(CNode& node_to, Peer& peer, std::chrono::mic
 
 void PeerManagerImpl::MaybeSendAddr(CNode& node, std::chrono::microseconds current_time)
 {
-    AssertLockHeld(node.cs_sendProcessing);
+    LOCK(node.cs_sendProcessing);
 
     // Nothing to do for non-address-relay peers
     if (!node.RelayAddrsWithConn()) return;
```

hebasto commented on src/net_processing.cpp:
```diff
@@ -4156,6 +4158,72 @@ void PeerManagerImpl::MaybeSendPing(CNode& node_to, Peer& peer)
+    }
+
+    // We sent an `addr` message to this peer recently. Nothing more to do.
+    if (current_time <= node.m_next_addr_send) return;
```
e3d6cbca83300e5ccc0fd8a1075ff52654aa76ff

Early return skips https://github.com/bitcoin/bitcoin/blob/e3d6cbca83300e5ccc0fd8a1075ff52654aa76ff/src/net_processing.cpp#L4221-L4224

That differs from original behavior.

hebasto commented on src/net_processing.cpp:
```diff
@@ -4156,6 +4158,72 @@ void PeerManagerImpl::MaybeSendPing(CNode& node_to, Peer& peer)
+                           node.vAddrToSend.end());
+
+    // No addr messages to send
+    if (node.vAddrToSend.empty()) return;
```
e3d6cbca83300e5ccc0fd8a1075ff52654aa76ff

Early return skips https://github.com/bitcoin/bitcoin/blob/e3d6cbca83300e5ccc0fd8a1075ff52654aa76ff/src/net_processing.cpp#L4221-L4224

That differs from original behavior.

jnewbery replied:
Thanks @hebasto! See https://github.com/bitcoin/bitcoin/pull/21236#issuecomment-809301197.

ccdle12 commented on src/net_processing.cpp:
```diff
@@ -4194,7 +4192,9 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
-    MaybeSendPing(*pto, *peer);
+    const auto current_time = GetTime<std::chrono::microseconds>();
+
+    MaybeSendPing(*pto, *peer, current_time);
```
super nit: `current_time` here is `const`. Not sure if the intention for `now` is to be immutable in `MaybeSendPing()`, but I think  `now` would still be mutable even though the outer scoped variable is `const` because of the param signature in `MaybeSendPing(..., std::chrono::microseconds now)`

jnewbery replied:
In general, there's no benefit to making pass by value parameters const. Passing by reference to const indicates to the caller that the parameter will not be mutated. When passing by value, a copy of the parameter is made, so the original object cannot be mutated.

There's very good documentation about how to pass parameters in the c++ core guidelines: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f16-for-in-parameters-pass-cheaply-copied-types-by-value-and-others-by-reference-to-const.

ccdle12 replied:
Thanks, apologies, I think I should have been more clear. Even though the `std::chrono` is passed by value, MaybeSendPing() treats it as mutable even though I don't think the method mutates the copy. So I was just thinking would it be useful/or maybe unecessary to make current_time in MaybeSendPing() also const.

jnewbery replied:
I understand. Yes, it is possible to mark the function parameters as `const` in the definition, which would prevent the copy from being mutated within the function body. You can even do this if the function declaration does not mark the parameter as `const`.

However, I haven't ever seen that style used in this project. It may be marginally useful, but it may also be misleading (eg see https://groups.google.com/u/4/g/comp.lang.c++.moderated/c/DEQTho8OCj8).

jnewbery committed:
[net] Change addr send times fields to be guarded by new mutex

jnewbery commented:
Thanks for the review @hebasto! Responding to your points:

> I see two calls of three are consolidated. What about the third one [...]

I was also counting the call inside `MaybeSendPing()`, which is called from `SendMessages()`. I've now removed the additional call to `GetTime()` as suggested, which required a duration cast.

> I believe that this commit is not a clean refactoring. To be precise, change behavior could be expected as even LOCK(cs_main); could take unpredictable time. But the scale of such changes seems insignificant to P2P, right?

Good observation! However, I think it's fine to treat time as constant within a call to `SendMessages()`, since it's observationally equivalent to any actors outside net_processing.

> The commit "[net processing] Extract addr send functionality into MaybeSendAddr()" (97e1013) allows to localize acquiring of the CNode::cs_sendProcessing mutex

The fact that you've raised this independently has made me reconsider the discussion here https://github.com/bitcoin/bitcoin/pull/21236#discussion_r583805388. It doesn't make sense for a mutex to  be guarding variables/functions across multiple layers, so I've moved the addr send time vars under their own mutex.

> Early return skips
> 
> ```
>  // we only send the big addr message once 
>  if (node.vAddrToSend.capacity() > 40) { 
>      node.vAddrToSend.shrink_to_fit(); 
>  } 
> ```
> 
> That differs from original behavior.

Another great observation! Although again, I don't think it matters. This logic is for when we've sent a response to the peer's `getaddr` message. In such cases, we'd expect to send up to 1000 addresses, and for the address filter to not yet be filled, so we wouldn't exit early.

hebasto commented:
`RecursiveMutex cs_sendProcessing` guards nothing. Let's burn it :)

hebasto commented:
pico-nit: I find myself more comfortable to reason about the code when the same variable is placed on the same position in similar logical operations. Consider
https://github.com/bitcoin/bitcoin/blob/b4d7d3def058e17d36cf7702e34c0a2fff45a5cf/src/net_processing.cpp#L4154-L4155 and https://github.com/bitcoin/bitcoin/blob/b4d7d3def058e17d36cf7702e34c0a2fff45a5cf/src/net_processing.cpp#L4173

Could `current_time` position be consistent?

jnewbery commented:
> pico-nit: I find myself more comfortable to reason about the code when the same variable is placed on the same position in similar logical operations.

I tend to agree and think that generally the variable should be placed on the left and the thing it's being compared to on the right. Others disagreed with me on previous PRs and told me to always place the smaller thing on the left. I think it's just a matter of taste :shrug:

hebasto reviewed (approved):
ACK 935d4889228e7e361c8b0020761fa0e08a55fb48, I have reviewed the code and it looks OK, I agree it can be merged.

sipa commented:
utACK 935d4889228e7e361c8b0020761fa0e08a55fb48

maflcko reviewed (commented):
review ACK 935d4889228e7e361c8b0020761fa0e08a55fb48 🐑

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 935d4889228e7e361c8b0020761fa0e08a55fb48 🐑
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUh7Hgv/dmVk2jwk66OpRVhWiCUITNoKMBCZFBA8+k63KsptFTEUV0oneUoiMVTm
Ftw+YVO60MZdNQjBet3CC3IyTUv098/H2+e9cODLgbRpRMJaLewzevxUy0o6RToQ
BzHalKxNnrGEUnJB/6zm4HzXWIPJWppY947BlLEuZ+HM7J8HgAmFVPZWa/VubsrN
YhY6rBbtHuJKGamsdE/LkNfNEMhP4dIOksl1WteXg2xoSeBHQeHL6nFLKFjJohbj
Y34ZSnxo0jJFpz+bSVlzelJh5EL7/G+fZzvE5vbOa39+5unZ2mFx70ftvGYgRK6V
qLfhnyeTMABXjs5bwU3HcjVGqa1ZsApIhY+sLs6lwSQmNAG2UjOxPcl7EDcXnZha
fV//g/jdU5qTyNQiriWrA0TEtwPm7X1lgJFYBgpfSAfnSIxbLk/uZWkPqFVVbt5P
Mh2PXN7lIqJ/irxKTBscAXqX4RrIbHXHVv5TaUvS3kI2igSLaKjUhj2Nqoea6UZ3
XBBbfMvT
=sVrE
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `2166a9ff7592f66162db55121e31cb0376f60455376bea865e5631ad552a2fc6  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401082166a9ff7592f66162db55121e31cb0376f60455376bea865e5631ad552a2fc6f0102cf489f84b3a2a6611e2babf6e0ee89508fff0103cf370e021bc498b55528e6ee8b32cca08f120431683da20373e484a106e2d272aa8fbefaba8b39e22a3006592c0755742509708f10460655a5ef008b752e9824c1824700083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010961ae5d19f889bafd6f48e75b61c509208f10460656837f008ddc32f51d23bdd6b0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010e729aca1dd4877f17e25494bd956163908f10460656837f0084facce2cfa9b9be70083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010fd84dd648cfeba09ce106563fe5205bc08f120f1d29a5e00da1ae44c9ec4ffdd6de340434e271be5f5e833b789fbc14223a50a08f10460656838f008ac15ef2b6c7824120083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>
